File: json-templates/src/main/java/nl/softcause/jsontemplates/expresionparser/ParseCursor.java
package nl.softcause.jsontemplates.expresionparser;

import java.util.Arrays;
import java.util.regex.Pattern;

import lombok.NonNull;

public class ParseCursor {

    private int cursorIndex = 0;
    private String text;
    private String full;

    ParseCursor(@NonNull String text) {
        this.text = text.trim();
        this.full = this.text;
    }

    private ParseCursor(String text, String full, int cursorIndex) {
        this.text = text;
        this.full = full;
        this.cursorIndex = cursorIndex;
    }

    public ParseCursor clone() {
        return new ParseCursor(text, full, cursorIndex);
    }

    boolean at(Pattern pattern) {
        var match = pattern.matcher(text);
        return match.find();
    }

    boolean at(String instruction) {
        return text.startsWith(instruction);
    }

    String read(Pattern pattern) {
        var match = pattern.matcher(text);
        if (!match.find()) {
            throw ParseException.expected(pattern).at(this);
        }
        var chunk = match.group(1);
        move(match.end());
        return chunk;
    }

    void read(String instruction) {
        if (!at(instruction)) {
            throw ParseException.expected(instruction).at(this);
        }
        move(instruction.length());
    }

    private void move(int length) {
        text = text.substring(length);
        cursorIndex += (length + (text.length() - text.trim().length()));
        text = text.trim();
    }

    boolean more() {
        return !text.isEmpty();
    }


    boolean at(String[] until) {
        return Arrays.stream(until).anyMatch(this::at);
    }

    @Override
    public String toString() {
        return text;
    }

    String getLeft() {
        return full.substring(cursorIndex);
    }

    String getFull() {
        return full;
    }

    int getIndex() {
        return cursorIndex;
    }
}

File: reference-api/src/main/js/model/ParseCursor.js
import _ from 'underscore';

export function ParseCursor(text){
    this.cursorIndex=0;
    this.text=text;
    this.full=text;
    this.lastBlock=null;

}

ParseCursor.prototype.more=function(){
    return this.text!=="";
}

ParseCursor.prototype.at=function(patternOrInstruction){
    if(_.isArray(patternOrInstruction)){
        return _.any(patternOrInstruction, i=>this.at(i));
    } else if(typeof(patternOrInstruction) === 'string'){
        return this.text.startsWith(patternOrInstruction);
    } else {
        return patternOrInstruction.test(this.text);
    }
}

ParseCursor.prototype.read=function(patternOrInstruction) {
    if(_.isArray(patternOrInstruction)){
        for(var i=0;i<patternOrInstruction.length;i++){
            if(this.at(patternOrInstruction[i])) {
                this.move(patternOrInstruction.length);
                return;
            }
            throw "Expected "+patternOrInstruction.join("|");
        }
    } else if(typeof(patternOrInstruction) === 'string') {
        if(!this.at(patternOrInstruction)) throw "Expected "+patternOrInstruction;
        this.move(patternOrInstruction.length);
    } else {
        const match = patternOrInstruction.exec(this.text);
        if (match.length===0) throw "Expected pattern "+patternOrInstruction;
        var chunk = match[0];//match.group(match[0]);
        // this.move(match.end());
        this.move(chunk.length);
        return chunk;
    }
}

ParseCursor.prototype.move=function(length){
    var start = this.cursorIndex;
    this.text=this.text.substring(length);
    this.cursorIndex+=(length+(this.text.length-this.text.trim().length));
    this.lastBlock = {
        start: start,
        end: this.cursorIndex
    }
    this.text=this.text.trim();
}

ParseCursor.prototype.getLastBlock=function(){
    return this.lastBlock;
}

ParseCursor.prototype.toString=function(){
    return this.text;
}

ParseCursor.prototype.getLeft=function(){
    return this.full.substring(this.cursorIndex);
}

ParseCursor.prototype.getFull=function(){
    return this.full;
}

ParseCursor.prototype.getIndex=function(){
    return this.cursorIndex;
}

export default ParseCursor;